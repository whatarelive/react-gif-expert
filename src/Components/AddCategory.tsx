// Librerias de React.
import React, {useState} from "react";
// Types.
import {AddCategoryProps} from "../Types/types";


export const AddCategory = ( { onNewCategory }: AddCategoryProps ) => {

    // Hook useState : permite dal un estado al texto del input.
    const [ inputValue, setInputValue ] = useState( String );

    // Funcion para extraer el texto del input y guardarlo en el useState.
    const onInputChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => setInputValue( target.value );

    // Funcion que enviar el valor de la nueva categoria al componente padre.
    const onSubmit = (event: React.FormEvent) => {
        event.preventDefault();

        const text = inputValue.trim();
        
        if( text.length > 0 ) {
            onNewCategory( text );
            setInputValue('');
        }
    };

    // Renderizado del componente.
    return (
        <form onSubmit={ onSubmit }>
            <input
                type="text"
                placeholder="Buscar Gifs"
                value={ inputValue }
                onChange={ onInputChange }
            />
        </form>
    );
};