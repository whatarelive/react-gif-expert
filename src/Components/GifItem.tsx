import { GifType } from "../Types/types"

export const GifItem = ( image: GifType ) => {

    return (
         <div className="card-item">          
            <img src={ image.url } alt={ image.title } /> 
            <p>{ image.title }</p>
        </div>      
    )
}
