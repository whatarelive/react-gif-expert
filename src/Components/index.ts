// Archivo de barril para todas las importaciones de la carpeta componentes.
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';