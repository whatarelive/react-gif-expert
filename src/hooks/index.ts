export * from './useCategory';
export * from './useFecthGifs';