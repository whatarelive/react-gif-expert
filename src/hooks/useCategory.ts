import { useState } from "react"

export const useCategory = ( newcategory: string ) => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState( [newcategory] )

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories])
        }
    }

    return {
        categories,
        onAddCategory
    }
}