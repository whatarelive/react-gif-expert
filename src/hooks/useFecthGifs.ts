// Hooks de React
import { useEffect, useState } from "react";
// API Helpers.
import { getGifs } from "../Helpers/getGifs";
// Types
import { GifType } from "../Types/types"


export const useFecthGifs = ( category: string ) => {

    // Hook useState : permite dal un estado del arreglo de elementos.
    const [ images, setImages ] = useState( Array<GifType> );

    const [ isLoading, setIsLoading ] = useState( true );

    // Funcion para recuperar el arreglo de gifs y guardarlos en el useState. 
    const getImages = async( category: string ) => {
        const newImages = await getGifs( category );
        setImages( newImages );
        setIsLoading( false );
    }

    // Hook UseEffect : permite que solo se llame a la funcion cuando halla un cambio en las categorias.
    useEffect( () => {
        getImages( category );

    }, [category])

    return {
        images,
        isLoading,
    }
}
