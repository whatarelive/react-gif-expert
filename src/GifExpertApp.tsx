// Librerias de React.
import {useState} from "react";
// Componentes hijos.
import { AddCategory, GifGrid } from "./Components";


export const GifExpertApp = () => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState(['One Punch']);

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if ( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories ]);
        }
    }

    // Funcion que crea el listado de elementos.
    const listItem = () => {
        return categories.map( ( category: string ) => {
           return <GifGrid key={ category } category={ category }/>;
        });
    }

    return (
        <>
            <h3>Gif Expert App</h3>
            <AddCategory onNewCategory={ onAddCategory }/>
            <ol>{ listItem() }</ol>
        </>
    );
};