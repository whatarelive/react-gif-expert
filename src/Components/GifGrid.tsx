// Custom Hook
import { useFecthGifs } from "../hooks/useFecthGifs";
// Componetes
import { GifItem } from "./index";
// Types.
import {GifGridProps} from "../Types/types";


export const GifGrid = ({ category }: GifGridProps ) => {

    const { images, isLoading } = useFecthGifs( category );
    
    // Renderizado del componente.
    return (
        <>
            <h3>{ category }</h3>

            {
                isLoading && ( <h2>Cargando.....</h2>)
            }

            <div className="card-grid"> 
            { 
                images.map( images => <GifItem key={ images.id } { ...images }/>)
            }
            </div> 
        </>
    );
};
