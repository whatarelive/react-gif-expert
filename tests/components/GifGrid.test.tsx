/* eslint-disable no-undef */
import React from "react";
import { render, screen } from "@testing-library/react";
import { GifGrid } from "../../src/Components";
import { useFecthGifs } from "../../src/hooks/useFecthGifs";

// Creamos un mock personalizado del customHook.
jest.mock("../../src/hooks/useFecthGifs");

describe('Pruebas en el componente <GifGrid/>', () => {

    // Variable auxiliar
    const category = "One Punch";

    // Test #1: Sobre la UI inicial del componente
    test('Snapshot del componente', () => { 

        // Definimos el valor de retorno de nuestro CustomHook
        (useFecthGifs as jest.Mock).mockReturnValue({
            images: [],
            isLoading: true
        })

        // Renderizamos el componente y guardamos el render.
        const { container } = render( <GifGrid category={ category }/>);

        // Creamos una captura del componente.
        expect( container ).toMatchSnapshot();
    })
    
    
    // Test #2: Sobre el mensaje de Carga.
    test('debe mostrar el loading inicialmente', () => {
        
        // Definimos el valor de retorno de nuestro CustomHook
        (useFecthGifs as jest.Mock).mockReturnValue({
            images: [],
            isLoading: true
        });

        // Renderizamos el componente
        render( <GifGrid category={ category }/>);

        // Esperamos que se muestre un elemento con el texto especificado.
        expect( screen.getByText('Cargando.....') );
        expect( screen.getByText( category ) );
    })


    // Test #3: Sobre la visibilidad de la lista de gifs.
    test('debe mostrar items cuando se cargan las imagenes useFetchGifs', () => { 

        // Creamos un arreglo auxiliar de elementos a mostrar.
        const gifs = [
            {
                id: 'ABC',
                title: 'Saitama',
                url: 'https://localhost/saitama.jpg'
            },
            {
                id: '123',
                title: 'Goku',
                url: 'https://localhost/goku.jpg'    
            }
        ];

        // Definimos el valor de retorno de nuestro CustomHook
        (useFecthGifs as jest.Mock).mockReturnValue({
            images: gifs,
            isLoading: false
        });
        
        // Renderizamos el componente.
        render( <GifGrid category={ category }/>);
        
        // Esoeramos que halla dos elementos de tipo imagen en el HTML.
        expect( screen.getAllByRole('img').length ).toBe(2);
    })
})