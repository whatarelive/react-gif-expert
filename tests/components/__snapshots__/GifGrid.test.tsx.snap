// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifGrid/> Snapshot del componente 1`] = `
<div>
  <h3>
    One Punch
  </h3>
  <h2>
    Cargando.....
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
